### ISSUE TITLE
Regression: geoarrow crate does not compile in release mode on 1.82

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#131960]
This issue is originally reported by @kylebarron at https://github.com/rust-lang/rust/issues/128887#issuecomment-2423159520

The `geoarrow` crate does not compile in release mode on rust 1.82.0, despite compiling fine in debug mode or in rust 1.81.0.

The code can be obtained by:
```sh
git clone https://github.com/geoarrow/geoarrow-rs
cd geoarrow-rs
git checkout 0b6715c6a56f0115f9078803fae945700713b22f
```

The following commands give compilation errors:
```sh
rustup run 1.82 cargo build --release
rustup run nightly cargo build --release
```

The following commands compile fine without errors:
```sh
rustup run 1.81 cargo build --release
rustup run 1.81 cargo build
rustup run 1.82 cargo build
rustup run nightly cargo build
RUSTFLAGS='-Zinline-mir=no' rustup run nightly cargo build --release
```

Note that this is issue is distinct from the previous 1.80-to-nightly geoarrow regression, which is fixed in #129714 and tested in #129757.

@rustbot labels regression-from-stable-to-stable

**Comment by saethlin** [comment]
> Note that this is issue is distinct from the previous 1.80-to-nightly geoarrow regression, which is fixed in [#129714](https://github.com/rust-lang/rust/pull/129714) and tested in [#129757](https://github.com/rust-lang/rust/pull/129757).

1. No, I'm pretty sure this is exactly the same problem. I see 128 recursive calls to `mir_callgraph_reachable::process` with -Ztreat-err-as-bug.

**Comment by saethlin** [comment]
We would also need a magic -3 adjustment on top of the existing recursion limit hack to make this compile. I could speculate but I don't think I understand why. I think it's pretty clear though that's the regression test we checked in doesn't cover the scenario that geoarrow is hitting.

**Comment by lqd** [comment]
Bisects to `nightly-2024-08-02`, seemingly #127159 within rollup #128461.

**Comment by saethlin** [comment]
cc @compiler-errors 

At the current time, I'm not working on any fix for this regression. I'm really not sure what a reasonable change would be; we can add some more magic numbers to the cycle avoidance recursion limit, but I'm not sure what number we can add that's reasonable. I suspect that whatever we do, there will be some type that someone can write that will run into this error anyway.

It's my understanding that this error can be fixed in the new trait solver, but that won't land for so long that it's kind of off topic for this issue. 

I feel like there should be a better way to write the cycle avoidance call graph code, but I don't know if this scenario can be avoided by that.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/channel/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23131960.20Regression.3A.20geoarrow.20crate.20does.20not.20compile.20in.20rele.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by theemathas** [comment]
Somewhat minimized reproduction code: [geoarrow-repro.zip](https://github.com/user-attachments/files/17464402/geoarrow-repro.zip)

The issue appears to be sensitive to the exact module paths and the exact dependencies used.

**Comment by theemathas** [comment]
This version of the code reproduces the issue on nightly, but not on stable. The difference is probably due to #130455, which makes MIR-inlining not depend on hashes.

<details>
<summary>Code</summary>

```rust
pub trait GeometryCollectionTrait: Sized {
    type T;

    type ItemType<'a>: 'a + GeometryTrait<T = Self::T>
    where
        Self: 'a;

    fn geometries(&self) -> GeometryCollectionIterator<'_, Self::T, Self::ItemType<'_>, Self> {
        unimplemented!()
    }
}
pub trait GeometryTrait {
    type T;

    type GeometryCollection<'a>: 'a + GeometryCollectionTrait<T = Self::T>
    where
        Self: 'a;

    fn as_type(&self) -> GeometryType<'_, Self::GeometryCollection<'_>>;
}

pub enum GeometryType<'a, GC>
where
    GC: GeometryCollectionTrait,
{
    GeometryCollection(&'a GC),
}

pub struct GeometryCollectionIterator<
    'a,
    T,
    ItemType: 'a + GeometryTrait<T = T>,
    G: GeometryCollectionTrait<T = T, ItemType<'a> = ItemType>,
> {
    _a: &'a G,
}

impl<
        'a,
        T,
        ItemType: 'a + GeometryTrait<T = T>,
        G: GeometryCollectionTrait<T = T, ItemType<'a> = ItemType>,
    > Iterator for GeometryCollectionIterator<'a, T, ItemType, G>
{
    type Item = ItemType;
    fn next(&mut self) -> Option<Self::Item> {
        unimplemented!()
    }
}
pub fn add_geometry(geometry: &impl GeometryTrait<T = f64>) {
    match geometry.as_type() {
        GeometryType::GeometryCollection(g) => add_geometry_collection(g),
    }
}

pub fn add_geometry_collection(geometry_collection: &impl GeometryCollectionTrait<T = f64>) {
    for geometry in geometry_collection.geometries() {
        add_geometry(&geometry);
    }
}

pub fn bounding_rect_geometry_collection(geom: &impl GeometryCollectionTrait<T = f64>) {
    add_geometry_collection(geom);
}
```

</details>

**Comment by theemathas** [comment]
Minimized:

```rust
pub trait One: Sized {
    type ItemOne;
    type ToTwo: Two<ItemTwo = Self::ItemOne>;
}
pub trait Two {
    type ItemTwo;
    type ToOne: One<ItemOne = Self::ItemTwo>;
}

trait OneExt {
    fn one_ext() {}
}
impl<T, C: One<ItemOne = T>> OneExt for C {}
// Using this instead makes it compile
// impl<C: One> OneExt for C {}

#[allow(unconditional_recursion)]
fn recurse<C: One>() {
    C::one_ext();
    recurse::<<C::ToTwo as Two>::ToOne>();
}

// This function is necessary to reproduce the bug.
pub fn call_recurse<C: One>() {
    recurse::<C>();
}
```

On stable rust 1.83.0, `cargo build` succeeds, but `cargo build --release` fails with an overflow.

**Comment by ianthetechie** [comment]
This seems to have reoccurred again in 1.90. See https://github.com/geoarrow/geoarrow-rs/issues/1339. The above minimized example seems to (still?) fail on 1.90 as well. Though some recent changes to the geoarrow crate unblocked builds for 1.89. As noted above and in the new issue report, debug builds work fine, but release builds on both stable and nightly will overflow.

**Comment by apiraino** [comment]
Visited during T-compiler triage meeting on https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202025-09-25/near/541470068.

We agree that probably the reasonable way forward is to fix this with an ad-hoc patch. Seems the new trait solver will obsolete these temporary solutions anyway. The new trait solver is in our [project goals](https://rust-lang.github.io/rust-project-goals/2025h2/next-solver.html) for the 2nd half of this year :-)

@rustbot label -I-compiler-nominated
